modules/develop/pages/data-transforms/how-transforms-work.adoc

```diff
@@ -0,0 +1,47 @@
+
+Wasm is executed within a runtime in the broker. Data transforms take an input topic and map it to an output topic, running a virtual machine (VM) on the same shard as every partition leader for a topic. Redpanda uses just-in-time (JIT) compilation, which compiles the function in memory, writes it to executable space, then runs the directly translated machine code. When a shard becomes the leader of a given partition on the input topic, Redpanda does the following:
+
+. Spins up a Wasm VM using the JIT-compiled WebAssembly module
+. Pushes records from the input partition into the Wasm VM
+. Writes the output, which may go to the same shard on the same machine, or to a different shard on the same machine, or to a different machine
+
+image::shared:wasm_architecture.png[Wasm architecture in Redpanda]
+
+The transform function processes every record produced to an input topic and returns zero or more records that are then produced to an output topic. Data transforms are applied to all partitions on an input topic. A record is processed after it has been successfully written to disk on the input topic. Because the transform happens in the background after the write finishes, the transform doesn't affect the original produced record and doesn't block writes to the input topic. 
+
+image::shared:wasm_flow.png[Flow of record transform]
+
+A new data transform processes the input topic from the latest offset. When a partition loses leadership, the broker issues an internal notification, and Redpanda shuts down the transform on that core. When another broker gains leadership of the partition, it also issues a notification, and the transform subsystem starts a processor and resumes from the last committed offset. This is why transforms have at-least-once semantics. 
+
+== Configure Wasm properties
+
+When you deploy a data transform, Redpanda pushes the generated `.wasm` files and metadata to the cluster, which then stores and replicates the bytecode and metadata using internal broker topics. Each shard of Redpanda runs its own copy of the transform function. There is at most one runtime for a transform for any shard, and memory for each function is reserved within the broker. This is achieved with the following cluster configuration properties: 
+
+- `wasm_per_core_memory_reservation`: Total amount of memory to reserve per shard for all Wasm VMs. Default = 20 MiB.
+- `wasm_per_function_memory_limit`: Amount of memory to reserve per instance of a Wasm VM. Default = 2 MiB.
+
+The maximum number of functions that can be deployed to a cluster is equal to `wasm_per_core_memory_reservation` / `wasm_per_function_memory_limit`. CPU time is dynamically allocated to the Wasm runtime and ensures that the code does not run forever and cannot block the broker from handling traffic or doing other work, such as Tiered Storage uploads.
```

Comment:
@rockwotj what happens when you try to deploy another transform once you've hit the limit? we should explain tht here @micheleRP

Reply:
Also the other , most important work we cannot block is produce and consume requests  @micheleRP

Reply:
Right now nothing except you get errors about unable to allocate memory for the VM and the transforms stay in error states.